### PR TITLE
Add new class API error and change API to return the same error struc…

### PR DESCRIPTION
**Pull request by rafinskipg**
- [x] Added new  ApiError class to handle api errors 
- [x] withApiHandler now handles the errors returned by this internal class and formats the response, also logs all the errors with good format
- [x] fetchJSON handles errors with this format and gets the right message
- [x] GQL requests uses ApiError to return formatted error info
- [x] Custom error500 page to show more info about the error
- [ ] Do not break app if GQL requests fails

image

### CODE REVIEW

File: modules/app/components/ErrorPage.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      <Box sx={{ maxWidth: '500px', margin: '0 auto', pt: '200px', paddingBottom: '200px' }}>
+        <Flex sx={{ justifyContent: 'center', alignItems: 'center' }}>
+          <Text sx={{ fontSize: 5, fontWeight: 'bold', mr: 2 }}>{statusCode}</Text> |{' '}
+          <Text sx={{ ml: 2 }}>{title}</Text>
+        </Flex>
+        <Box pt={3}>{children}</Box>
+        <Box pt={3}>
+          <Text sx={{ textAlign: 'center' }}>
+            For more information or help contact the Development and UX core unit at{' '}
+            <ExternalLink href="https://discord.gg/GHcFMdKden" title="Discord">
```
Comment (adamgoth) [inline modules/app/components/ErrorPage.tsx:30]: `For more information or help, please contact the Development & UX core unit on Discord` reads a little better.

File: modules/mkr/helpers/getMKRVotingWeight.ts [hunk 2]
```diff
@@ -70,7 +70,7 @@ export async function getMKRVotingWeight(
   // otherwise, not proxy or delegate, get connected wallet balances
   const walletBalanceHot = await contracts.mkr.balanceOf(address);
   const chiefBalanceHot = await contracts.chief.deposits(address);
-
+  console.log('eeeeee', chiefBalanceHot.toNumber());
```
Comment (adamgoth) [inline modules/mkr/helpers/getMKRVotingWeight.ts:73]: ✂️

File: pages/address/[address]/index.tsx [hunk 3]
```diff
@@ -110,8 +110,18 @@ export default function AddressPage(): JSX.Element {
...
+  if (error && error.status === 404) {
+    return (
+      <PrimaryLayout>
+        <ErrorPage statusCode={404} title="This address does not exist" />
+      </PrimaryLayout>
+    );
+  } else if (error) {
+    return (
+      <PrimaryLayout>
+        <ErrorPage statusCode={500} title="Error fetching address information" />
```
Comment (adamgoth) [inline pages/address/[address]/index.tsx:122]: Are we able to pass `error.status` instead of hardcoding 500?

File: pages/api/address/stats.ts [hunk 4]
```diff
@@ -84,9 +85,7 @@ export default withApiHandler(async (req: NextApiRequest, res: NextApiResponse)
       : (req.query.address as string[]);
 
   if (!req.query.address) {
-    return res.status(400).json({
-      error: 'Missing address'
-    });
+    throw new ApiError('Address stats, missing address', 401, 'Missing address');
```
Comment (adamgoth) [inline pages/api/address/stats.ts:88]: 401 is typically used for `Unauthorized` where 400 is for `bad request`. Wouldn't 400 make more sense here?

**Review by adamgoth — COMMENTED** [review]
Looks good. Just left a couple comments about small things.

File: pages/api/address/stats.ts (hunk 4 above)
Comment (rafinskipg) [reply pages/api/address/stats.ts:88]: yes! typo

**Pull request merged**